Feature/Adding RoleAdminChanged event in AccessControl

julianmrodri opened the pull request:
<!-- 0. 🎉 Thank you for submitting a PR! -->

<!-- 1. Does this close any open issues? Please list them below. -->
#2210  

<!-- Keep in mind that new features have a better chance of being merged fast if
they were first discussed and designed with the maintainers. If there is no
corresponding issue, please consider opening one for discussion first! -->

Fixes #2210

<!-- 2. Describe the changes introduced in this pull request. -->
<!--    Include any context necessary for understanding the PR's purpose. -->

Adding a new event named "RoleAdminChanged" when the adminRole of a Role is changed. This event is emitted from with the _setRoleAdmin function.

<!-- 3. Before submitting, please make sure that you have:
  - reviewed the OpenZeppelin Contributor Guidelines
    (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/CONTRIBUTING.md),
  - added tests where applicable to test new functionality,
  - made sure that your contracts are well-documented,
  - run the Solidity linter (`npm run lint:sol`) and fixed any issues,
  - run the JS linter and fixed any issues (`npm run lint:fix`), and
  - updated the changelog, if applicable.
-->

julianmrodri commented:
I see 2 potential changes worth discussing:

- I put the Declaration of the new event before the existing ones, we can move it to the after the other two events.
- Should we only trigger the event in the case the new Role Admin is different than the existing Role Admin? (this would require to add a validation)

Looking forward for your comments!
.

julianmrodri commented:
@nventuro @frangio  Would really appreciate your inputs so we can put this puppy to bed! Thanks for your time!

nventuro commented on contracts/access/AccessControl.sol:
```diff
@@ -52,6 +52,11 @@ abstract contract AccessControl is Context {
     bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
 
+    /**
+     * @dev Emitted when `adminRole` is set as ``role``'s admin role.
```
```suggestion
     * @dev Emitted when `adminRole` is set as ``role``'s admin role.
     *
     * `DEFAULT_ADMIN_ROLE` is the starting admin for all roles, despite
     * {RoleAdminChanged} not being emitted signaling this.
```

frangio commented:
I agree we should include the current admin role in the event as well, similar to the `OwnershipTransferred` event. And I think we should always emit it.

julianmrodri commented:
Sounds good! Thanks! Will implement the change to add the existing Admin Role and let you know.

julianmrodri commented:
@nventuro @frangio  I have implemented the changes you suggested. 

- I added the "_previousAdminRole_" to the event. I followed the same naming you are using in `OwnershipTransferred`.  I had to emit the event before updating the value now (the same is being done in 'OwnershipTransferred'). 
- Regarding the tests I had to break one statement into 3 lines due to **lint** issues (max length). 

Let me know what you think thanks for your support!

julianmrodri commented:
Thanks for your time and patience! for sure!
